src/components/motor/server.cpp

```diff
@@ -0,0 +1,232 @@
+    motor->stop_motor();
+
+    return ::grpc::Status();
+}
+
+::grpc::Status MotorServer::IsPowered(::grpc::ServerContext* context,
+                                      const ::viam::component::motor::v1::IsPoweredRequest* request,
+                                      ::viam::component::motor::v1::IsPoweredResponse* response) {
+    if (request == nullptr) {
+        return ::grpc::Status(::grpc::StatusCode::INVALID_ARGUMENT,
+                              "Called [Motor::IsPowered] without a request");
+    };
+
+    std::shared_ptr<ResourceBase> rb = sub_svc->resource(request->name());
+    if (rb == nullptr) {
+        return grpc::Status(grpc::UNKNOWN, "resource not found: " + request->name());
+    }
+
+    std::shared_ptr<Motor> motor = std::dynamic_pointer_cast<Motor>(rb);
+
+    Motor::power_status result = motor->get_power_status();
+    response->set_is_on(result.is_on);
+    response->set_power_pct(result.power_pct);
+
+    return ::grpc::Status();
+}
+
+::grpc::Status MotorServer::IsMoving(::grpc::ServerContext* context,
+                                     const ::viam::component::motor::v1::IsMovingRequest* request,
+                                     ::viam::component::motor::v1::IsMovingResponse* response) {
+    if (request == nullptr) {
+        return ::grpc::Status(::grpc::StatusCode::INVALID_ARGUMENT,
+                              "Called [Motor::IsMoving] without a request");
+    };
+
+    std::shared_ptr<ResourceBase> rb = sub_svc->resource(request->name());
+    if (rb == nullptr) {
+        return grpc::Status(grpc::UNKNOWN, "resource not found: " + request->name());
+    }
+
+    std::shared_ptr<Motor> motor = std::dynamic_pointer_cast<Motor>(rb);
+
+    bool result = motor->is_moving();
+    response->set_is_moving(result);
+
+    return ::grpc::Status();
+}
+
+::grpc::Status MotorServer::DoCommand(grpc::ServerContext* context,
+                                      const viam::common::v1::DoCommandRequest* request,
+                                      viam::common::v1::DoCommandResponse* response) {
+    if (request == nullptr) {
+        return ::grpc::Status(::grpc::StatusCode::INVALID_ARGUMENT,
+                              "Called [Motor::DoCommand] without a request");
+    };
+
+    std::shared_ptr<ResourceBase> rb = sub_svc->resource(request->name());
+    if (rb == nullptr) {
+        return grpc::Status(grpc::UNKNOWN, "resource not found: " + request->name());
+    }
+
+    std::shared_ptr<Motor> motor = std::dynamic_pointer_cast<Motor>(rb);
+    AttributeMap result = motor->do_command(struct_to_map(request->command()));
+
+    *response->mutable_result() = map_to_struct(result);
+    /* response->mutable_result()->CopyFrom(map_to_struct(result)); */
+
+    return ::grpc::Status();
+}
+
+void MotorServer::register_server(std::shared_ptr<Server> server) {
+    viam::component::motor::v1::MotorService::Service* motor =
+        static_cast<viam::component::motor::v1::MotorService::Service*>(this);
```

Comment:
I think I disagree. Using `static_cast` here is correct because you know that `MotorServer` derives from that type. There is no way the cast can fail. For that matter, I don't even see why the cast is needed. I'd expect `server->register_service(this)` to work.

Reply:
Removed `static_cast` without regressions.